README.md
[![Build Status](https://travis-ci.org/lucid-services/serviser-template.svg?branch=master)](https://travis-ci.org/lucid-services/serviser-template)  

Rapid start of a new [serviser](https://github.com/lucid-services/serviser) based project.  

##### CLI interface

```bash
> serviser-template -h
node_modules/.bin/serviser-template <command> [options]

Commands:
  init       Creates a new serviser project

Options:
  --verbose, -v      Dumps more info to stdout                  [count] [default: 1]
  -h, --help         Show help                                  [boolean]
```


### Tests

```bash
> npm test
```


CHANGELOG.md
## 0.3.0

* [CHANGED] - renamed project to `serviser-template`
* [ADDED] - new `serviser` plugin options in dependencies installation stage

## 0.2.0

* [ADDED] - `serviser-shell` plugin support
* [ADDED] - npm test initialization before/after hooks
* [CHANGED] - renamed generated test/test.js -> test/index.js

## 0.1.0

* [ADDED] - initial release


test/cli.js
'use strict';
const _            = require('lodash');
const http         = require('http');
const Promise      = require('bluebird');
const os           = require('os');
const fs           = Promise.promisifyAll(require('fs'));
const chai         = require('chai');
const tmp          = require('tmp');
const path         = require('path');
const yargs        = require('yargs');
const json5        = require('json5');
const childProcess = require('child_process');

// adds .json5 loader require.extension
require('json5/lib/require');

Object.defineProperty(global, 'Promise', {
    configurable: false,
    writable: false,
    value: Promise
});

const cliInterface = require('../index.js');

chai.should();

describe('yargs cli interface', function() {
    it('should return argv builder object', function() {
        let argvBuilder = cliInterface(yargs, false);
        argvBuilder.should.be.an('object');
        argvBuilder.should.have.property('argv');
        argvBuilder.should.have.property('coerce').that.is.a('function');
        argvBuilder.should.have.property('command').that.is.a('function');
    });
});

describe('bin/serviser-template', function() {
    before('Generating serviser skeleton', function(done) {
        this.slow(30000);

        let self = this;
        tmp.setGracefulCleanup();
        this.tmpDir = tmp.dirSync({unsafeCleanup: true});

        this.questions = [
            `? The project name (${path.basename(this.tmpDir.name)})`,
            '? Version (1.0.0)',
            '? Description',
            '? Author',
            '? Keywords (comma separated)',
            '? License (Use arrow keys or type to search)',
            '? Dependencies',
            '? npm test cmd',
            '? Apps',
            '? Service Host (http://127.0.0.1)',
            '? public app port',
            '? public-doc app port',
            '? cli app port',
        ];

        let proc = childProcess.spawn('node', [
            path.resolve(__dirname + '/../bin/serviser-template.js'),
            'init'
        ], {cwd: this.tmpDir.name});

        proc.stdout.on('data', function(chunk) {
            //when quetion prompt is activated, accept default value
            let questionCandidate = chunk.toString();
            process.stdout.write('      ' + questionCandidate);
            for (let i = 0, len = self.questions.length; i < len; i++) {
                if (   (self.questions[i] instanceof RegExp
                    && questionCandidate.match(self.questions[i]))
                    || (questionCandidate.indexOf(self.questions[i]) !== -1)
                ) {
                    self.questions.splice(i, 1);
                    proc.stdin.write(os.EOL);
                    return;
                }
            }
        });

        proc.on('error', function(e) {
            done(e);
        });

        proc.stderr.on('data', function(chunk) {
            console.error(chunk.toString())
        });

        proc.on('close', function(code) {
            if (code !== 0) {
                return done(new Error(`Exited with status code: ${code}`));
            }
            done();
        });
    });

    it('should generate runnable serviser project skeleton with default option values', function(done) {
        let config = require(path.resolve(this.tmpDir.name + '/config/config.js'));
        let cliPort = config.apps.cli.listen;
        let proc = childProcess.spawn('npm', ['start'], {cwd: this.tmpDir.name});

        proc.on('error', function(e) {
            done(e);
        });

        proc.stdout.on('data', function(chunk) {
            if (chunk.toString().indexOf('cli app listening on port') !== -1) {
                checkServiceIntegrity(cliPort, function(err) {
                    proc.removeAllListeners('close');
                    proc.kill();
                    done(err);
                });
            }
        });

        proc.stderr.on('data', function(chunk) {
            console.error(chunk.toString());
        });

        proc.on('close', function(code) {
            return done(new Error(`Service process exited unexpectedly with status code: ${code}`));
        });

    });
});

function checkServiceIntegrity(port, done) {
    let req = http.get({
        host: '127.0.0.1',
        port: port,
        path: '/api/v1.0/integrity',
        method: 'GET'
    }, function(res) {
        let data = '';
        res.setEncoding('utf8');
        res.on('data', function (chunk) {
            data += chunk.toString();
        });

        res.on('end', function() {
            if (res.statusCode !== 200 && res.statusCode !== 302) {
                return done(
                    new Error(`Response status code: ${res.statusCode}. ${data}`)
                );
            }

            return done(null, JSON.parse(data));
        });
    });

    req.once('error', done);
    req.end();
}


test/template.js
'use strict';
const _              = require('lodash');
const Promise        = require('bluebird');
const fs             = Promise.promisifyAll(require('fs'));
const sinon          = require('sinon');
const chai           = require('chai');
const sinonChai      = require("sinon-chai");
const chaiAsPromised = require('chai-as-promised');
const tmp            = require('tmp');
const path           = require('path');

Object.defineProperty(global, 'Promise', {
    configurable: false,
    writable: false,
    value: Promise
});

const Template = require('../lib/template.js');
const userInput = require('./userInput.json');

//this makes sinon-as-promised available in sinon:
require('sinon-as-promised', Promise);

const expect = chai.expect;

chai.use(sinonChai);
chai.should();
chai.use(chaiAsPromised);

describe('Template', function() {
    beforeEach(function() {
        tmp.setGracefulCleanup();
        this.tmpDir = tmp.dirSync({unsafeCleanup: true});

        this.template = new Template({cwd: this.tmpDir.name});

        this.getAnswersStub = sinon.stub(Template.prototype, '_getAnswers');
    });

    afterEach(function() {
        this.getAnswersStub.restore();
    });

    it('should generate serviser project files', function() {
        let self = this;
        this.getAnswersStub.resolves(_.cloneDeep(userInput));

        let files = [
            '/package.json',
            '/config/config.js',
            '/index.js',
            '/LICENSE',
            '/CHANGELOG.md',
            '/README.md',
            '/lib/app.js',
            '/.gitignore',
            '/.npmignore',
            '/test/index.js'
        ];

        return this.template.initCmd({
            verbose: 0,
            npm: false
        }).bind(this).then(function() {
            return Promise.map(files, function(p) {
                return fs.statAsync(path.resolve(self.template.options.cwd + p))
                    .then(function(stat) {
                        stat.isFile().should.be.equal(true, `${p} should be a file`);
                    });
            });
        });
    });

    it('should populate the project`s LICENSE file', function() {
        let self = this;
        let _userInput = _.cloneDeep(userInput);
        _userInput.license = "MIT License";
        this.getAnswersStub.resolves(_userInput);

        return this.template.initCmd({
            verbose: 0,
            npm: false
        }).bind(this).then(function() {
            return fs.readFileAsync(path.resolve(self.template.options.cwd + '/LICENSE'))
                .then(function(data) {
                    let mitLicense = fs.readFileSync(
                        path.resolve(__dirname + '/mitLicense.txt')
                    );

                    data.toString().should.be.equal(mitLicense.toString());
                });
        });
    });

    it('should generate lib/database/couchbase.js connector module', function() {
        let self = this;
        let _userInput = _.cloneDeep(userInput);
        _userInput._dependencies.push('serviser-couchbase');
        _userInput._config = {
            storage: {
                couchbase: {
                    host: 'couchbase://127.0.0.1',
                    buckets: {
                        default: {bucket: 'default'}
                    }
                }
            }
        };

        this.getAnswersStub.resolves(_userInput);

        return this.template.initCmd({
            verbose: 0,
            npm: false
        }).bind(this).then(function() {
            let p = path.resolve(
                self.template.options.cwd + '/lib/database/couchbase.js'
            );

            return fs.statAsync(p)
                .then(function(stat) {
                    stat.isFile().should.be.equal(true);
                });
        });
    });
});


index.js
const Template = require('./lib/template.js');

module.exports          = cliInterface;
module.exports.Template = Template;

let template = module.exports.template = new Template();

function cliInterface(yargs, strict) {

    template.CLI = true;

    return yargs
    .usage('$0 <command> [options]')
    .command(['init'], 'Initializes a new serviser project', {
        npm: {
            describe: 'Whether to run npm install after a project is generated',
            type: 'boolean',
            default: true
        }
    }, template.initCmd.bind(template))
    .option('verbose', {
        alias: 'v',
        describe: 'Dumps more info to stdout',
        default: 1,
        count: true,
        global: true,
        type: 'boolean'
    })
    .example('$0 init --no-npm', '# Generate project skeleton and skip installation of npm dependencies')
    .strict(typeof strict === 'boolean' ? strict : true);
}


lib/questions.js
const path = require('path');
const _    = require('lodash');
const url  = require('url');

const plugins  = require('./plugins.json');
const licenses = require('./licenses.json');


/**
 * collection of question getters (functions)
 * each function takes answers from previous set of questions and returns
 * a new question set (Array<Object>)
 */
let questions = module.exports = [];

questions.push(function() {
    return [
        {
            name: 'name',
            message: 'The project name',
            default: path.basename(process.cwd())
        },
        {
            name: 'version',
            message: 'Version',
            default: '1.0.0'
        },
        {
            name: 'description',
            message: 'Description',
        },
        {
            name: 'author',
            message: 'Author',
        },
        {
            name: 'keywords',
            message: 'Keywords (comma separated)',
            filter: function(keywords) {
                return keywords.split(',').map(function(keyword) {
                    return keyword.trim();
                });
            }
        },
        {
            type: 'autocomplete',
            name: 'license',
            pageSize: 10,
            message: 'License',
            source: searchLicense
        },
        {
            type: 'checkbox',
            name: '_dependencies',
            message: 'Dependencies',
            pageSize: 17,
            default: _.reduce(plugins, function(out, options, name) {
                if (options.default) {
                    out.push(name);
                }
                return out;
            }, []),
            choices: Object.keys(plugins)
        },
        {
            name: 'scripts.test',
            message: 'npm test cmd',
            default: function(argv) {
                if (argv._dependencies.indexOf('mocha')) {
                    return "mocha --ui bdd --colors --check-leaks -t 5000 --reporter spec 'test/**/*.js'";
                }
            }
        },
        {
            name: '_config._apps',
            message: 'Apps (Comma separated list of names of apps which will be created)',
            default: 'public',
            filter: function(apps, argv) {
                apps = apps.split(',').reduce(function(out, app) {
                    app =  app.trim();

                    out.push(app);

                    if (~argv._dependencies.indexOf('serviser-doc')) {
                        out.push(app + '-doc');
                    }

                    return out;
                }, []);

                if (~argv._dependencies.indexOf('serviser-cli')) {
                    apps.push('cli');
                }

                deriveAppTypes(argv, apps);

                return apps;
            }
        }
    ];
});

questions.push(function(argv) {
    let q = [
        {
            name: '_config._host',
            message: 'Service Host',
            default: 'http://127.0.0.1',
            filter: function(host) {
                let data = url.parse(host);
                if (!data.protocol) {
                    return 'http://' + host;
                }
                return host;
            }
        }
    ];

    argv._config._apps.forEach(function(app, index) {
        q.push({
            name: `_config.listen.${app}.port`,
            message: `${app} app port`,
            default: getPort()
        });
    });

    if (~argv._dependencies.indexOf('serviser-couchbase')) {
        q.push({
            name: '_config.storage.couchbase.host',
            message: 'Couchbase Host',
            default: 'couchbase://127.0.0.1',
            filter: function(host) {
                let data = url.parse(host);
                if (!data.protocol) {
                    return 'couchbase://' + host;
                } else {
                    return 'couchbase://' + data.host;
                }
            }
        });

        q.push({
            name: '_config.storage.couchbase.buckets',
            message: 'Buckets (comma separated list)',
            default: 'default',
            filter: function(buckets) {
                return buckets.split(',').reduce(function(out, bucket) {
                    bucket = bucket.trim();
                    out[bucket] = {bucket: bucket};
                    return out;
                }, {});
            }
        });
    }

    if (~argv._dependencies.indexOf('serviser-sequelize')) {
        q.push({
            name: '_config._sqlProvider',
            type: 'list',
            message: 'SQL Provider',
            default: 'postgres',
            choices: ['postgres', 'mysql']
        });

        q.push({
            name: '_config._sqlHost',
            message: function(argv) {
                return _.upperFirst(argv._config._sqlProvider) + ' Host';
            },
            default: '127.0.0.1'
        });

        q.push({
            name: '_config._sqlDatabase',
            message: function(argv) {
                return _.upperFirst(argv._config._sqlProvider) +
                    ' Database';
            },
            default: argv.name,
            filter: function(database) {
                return database.trim();
            }
        });

        q.push({
            name: '_config._sqlUsername',
            message: function(argv) {
                return _.upperFirst(argv._config._sqlProvider) +
                    ' Database username';
            }
        });

        q.push({
            name: '_config._sqlPassword',
            message: function(argv) {
                return _.upperFirst(argv._config._sqlProvider) +
                    ' Database password';
            }
        });
    }

    return q;
});

function searchLicense(answers, input) {
    return new Promise(function(resolve) {
        resolve(_.map(licenses.filter(filter(input)), 'name'));
    });
}

function filter(input) {
    return function(val) {
        if (!input) {
            return true;
        }
        return new RegExp(input, 'i').exec(val.name) !== null;
    };
}

function getPort() {
    let port = Math.floor(Math.random() * (4000 - 3000 + 1)) + 3000;

    if (!getPort._ports) {
        getPort._ports = [];
    }

    if (~getPort._ports.indexOf(port)) {
        return getPort();
    } else {
        return port + '';
    }
}

/**
 * @param {Object} argv
 * @param {Array<String>} apps
 * @return {undefined}
 */
function deriveAppTypes(argv, apps) {
    apps.forEach(function(app) {
        let type;

        if (~['s2s', 'private', 'cli'].indexOf(app)) {
            type = 'private'
        } else if (~['public'].indexOf(app)) {
            type = 'public';
        }

        if (type) {
            _.set(argv, ['_config', 'listen', app, 'type'], type);
        }
    });
}


lib/template.js
const https        = require('https');
const Promise      = require('bluebird');
const path         = require('path');
const fs           = Promise.promisifyAll(require('fs'));
const Prompt       = require('inquirer');
const mustache     = require('mustache');
const _            = require('lodash');
const json5        = require('json5');
const childProcess = require('child_process');
const chalk        = require('chalk');

const questions  = require('./questions.js');
const plugins    = require('./plugins.json');
const licenses   = require('./licenses.json');

Prompt.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'));

module.exports = Template;
module.exports.Template = Template;

/**
 * @public
 * @param {Object} options
 * @param {String} options.cwd=process.cwd(
 *
 * @constructor
 */
function Template(options) {
    options = options || {};
    this.CLI = null;

    if (!options.cwd) {
        options.cwd = process.cwd();
    }

    this.options = options;

    this.defaults = {
        scripts: {
            start: "./node_modules/.bin/serviser run",
        },
        private: true,
        main: 'index.js',
        files: [
            'CHANGELOG.md',
            'README.md',
            'LICENSE',
            'index.js',
            'lib',
            'bin'
        ],
        engines: {
            node: `>=${process.version.slice(1)}`
        },
        contributors: [],
        dependencies: {},
        devDependencies: {},
        peerDependencies: {}
    };
}

/**
 * @param {Object} argv
 * @return {Promise}
 */
Template.prototype.initCmd = Promise.method(function(argv) {
    let pkg, config, dependencies;

    return this._getAnswers().bind(this).then(function(answers) {
        dependencies = Template._normalizeDependencies(answers._dependencies, answers);
        config = Template._normalizeConfig(answers._config);
        pkg = Template._extractNpmPackage(answers);

        if (pkg.license !== 'None') {
            return Template._getLicense(pkg.license);
        }
        return null;
    }).then(function(license) {
        return this.create(pkg, config, dependencies, license, argv.verbose);
    }).then(function() {
        if (!argv.npm) {
            return null;
        }

        return Template._npmInstall(
            dependencies.dependencies,
            [],
            this.options.cwd,
            argv.verbose
        );
    }).then(function() {
        if (!argv.npm) {
            return null;
        }

        return Template._npmInstall(dependencies.devDependencies,
            ['--save-dev'],
            this.options.cwd,
            argv.verbose
        );
    });
});


/**
 * @param {Object} package
 * @param {Object} config
 * @param {Object} dependencies
 * @param {Object} license
 * @param {String} license.body
 * @param {Integer} verbosity - logging level
 *
 * @return {Promise}
 */
Template.prototype.create = function(package, config, dependencies, license, verbosity) {
    let self = this;
    let appContext = _.reduce(config.apps, function(out, val, app) {
        if (!~['cli'].indexOf(app)) {
            out.apps.push(app);
        }
        return out;
    }, {apps: []});

    let directories = [
        '/lib',
        '/lib/routes',
        '/lib/routes/v1.0',
        '/config',
        '/test',
    ];

    let templates = [
        {
            path: '/package.json',
            data: this._render('package', package)
        },
        {
            path: '/config/config.js',
            data: this._render('config', config, {json5: true})
        },
        {
            path: '/index.js',
            data: this._render('index', {
                package: Object.assign({}, package, {
                    dependencies: dependencies.dependencies,
                    devDependencies: dependencies.devDependencies
                }),
                config: config
            })
        },
        {
            path: '/LICENSE',
            data: this._render('LICENSE', {
                license: license || ''
            })
        },
        {
            path: '/CHANGELOG.md',
            data: this._render('CHANGELOG')
        },
        {
            path: '/README.md',
            data: this._render('README')
        },
        {
            path: '/lib/app.js',
            data: this._render('app', appContext)
        },
        {
            path: '/.gitignore',
            data: this._render('gitignore')
        },
        {
            path: '/.npmignore',
            data: this._render('npmignore')
        },
        {
            path: '/test/index.js',
            data: this._render('test')
        },
    ];

    if (_.has(config, 'storage.couchbase')) {
        directories.push('/lib/database');
        directories.push('/lib/models');
        directories.push('/lib/models/odm');
        templates.push({
            path: '/lib/database/couchbase.js',
            data: this._render('couchbase')
        });
    }

    appContext.apps.forEach(function(appName) {
        directories.push(`/lib/routes/v1.0/${appName}`);
        directories.push(`/lib/routes/v1.0/${appName}/example`);
        directories.push(`/lib/routes/v1.0/${appName}/example/routes`);

        templates.push({
            path: `/lib/routes/v1.0/${appName}/example/router.js`,
            data: this._render('router', {app: appName})
        });
        templates.push({
            path: `/lib/routes/v1.0/${appName}/example/routes/get.js`,
            data: this._render('route')
        });
    }, this);

    //create directories
    directories.forEach(function(dir) {
        return fs.mkdirSync(path.resolve(this.options.cwd + dir));
    }, this);

    return Promise.map(templates, function(template) {
        if (verbosity) {
            console.info(
                chalk.cyan('[INFO]') +
                ` Creating ${template.path.slice(1)}`
            );
        }
        let p = path.resolve(self.options.cwd + template.path);
        return fs.writeFileAsync(p, template.data);
    });
};

/**
 * @private
 * @return {Object}
 */
Template.prototype._getAnswers = Promise.method(function() {
    return Promise.reduce(questions, function(answers, questionGetter) {
        return Prompt.prompt(questionGetter(answers)).then(function(_answers) {
            return _.merge(answers, _answers);
        });
    }, this.defaults);
});

/**
 * @param {String} name
 * @param {Object} data - template context data
 * @param {Object} options
 * @param {Bolean} options.json5
 *
 * @return {String}
 */
Template.prototype._render = function(name, data, options) {
    options = options || {};

    const tmpl = fs.readFileSync(
        path.resolve(__dirname + `/templates/${name}.mustache`)
    );

    let context = _.reduce(data, function(out, value, key) {
        out['_' + key] = value;

        if (options.json5) {
            out[key] = new Function(`return \`${json5.stringify(value, null, 4)}\`;`);
        } else {
            out[key] = new Function(`return JSON.stringify(this['_${key}'], null, 4);`);
        }

        if (value && value._transform instanceof Function) {
            //TODO _fn leaks as its used inside template
            _fn = out[key];
            out[key] = function() {
                return value._transform.call(this, _fn.call(this));
            };
        }
        return out;
    }, {});

    return mustache.render(tmpl.toString(), context);
};

/**
 * @param {Array<String>} dependencies - optional user defined dependencies
 * @param {Object}        answers
 * @param {Object}        answers._config
 * @return {Object}
 */
Template._normalizeDependencies = function normalizeProjectDependencies(dependencies, answers) {
    dependencies = dependencies || [];
    let out = {
        dependencies: {
            'serviser': '*',
            bluebird: '*',
            lodash: '*'
        },
        devDependencies: {}
    };

    switch (answers._config._sqlProvider) {
        case 'postgres':
            out.dependencies.pg = '^4.5.0';
            break;
        case 'mysql':
            out.dependencies.mysql = '*';
            break;
    }

    return dependencies.reduce(function(out, dep) {
        if (plugins.hasOwnProperty(dep)) {
            if (plugins[dep].dev) {
                out.devDependencies[dep] = '*';
            } else {
                out.dependencies[dep] = '*';
            }
        }

        return out;
    }, out);
}

/**
 * @param {Object} config
 * @return {Object}
 */
Template._normalizeConfig = function normalizeProjectConfig(config) {
    const out = {
        listen: config.listen,
        storage: config.storage || {},
        apps: {}
    };

    /*
     * JSON.stringify always wraps string values in quotes, we need to get
     * rid of them in specific cases as we want to reference variables defined
     * in the template
     */
    Object.defineProperty(out.apps, '_transform', {
        value: function(str) {
            return str.replace(/bodyParser: "bodyParser"/g, 'bodyParser: bodyParser')
                .replace(/response: "response"/g, 'response: response');
        }
    })

    //normalize config.storage.mysql/postgres
    if (config._sqlProvider) {
        out.storage[config._sqlProvider] = {
            host: config._sqlHost,
            ssl: false,
            databases: {
                main: {
                    db: config._sqlDatabase,
                    username: config._sqlUsername,
                    password: config._sqlPassword,
                }
            }
        };
    }

    //normalize config.apps
    (config._apps || []).filter(function(appName) {
        return !~appName.indexOf('-doc');
    }).forEach(function(appName) {

        out.apps[appName] = {
            baseUrl: config._host + ':' + config.listen[appName].port,
            listen: config.listen[appName].port,
            stopOnError: true,
            bodyParser: 'bodyParser',//variable reference (defined in template)
            response: 'response', //variable reference (defined in template)
        };

        if (   config._apps instanceof Array
            && config._apps.indexOf(`${appName}-doc`) !== -1
        ) {
            out.apps[appName].doc = {
                baseUrl: config._host + ':' + config.listen[appName+'-doc'].port,
                listen: config.listen[appName+'-doc'].port,
                title: 'Docs',
                stopOnError: true
            };
        }
    });

    return out;
}

/**
 * @param {Object} answers
 * @return {Object}
 */
Template._extractNpmPackage = function extractNpmPackage(answers) {
    return _.reduce(answers, function(out, value, key) {
        if (!key.match(/^_.+$/)) {
            out[key] = _.cloneDeep(value);
        }

        return out;
    }, {});
}

/**
 * @param {String} name
 * @return {String}
 */
Template._getLicense = function getLicense(name) {
    const license = _.find(licenses, ['name', name]);

    return new Promise(function(resolve, reject) {
        let req = https.get({
            host: 'api.github.com',
            path: `/licenses/${license.key}`,
            headers: {
                Accept: 'application/vnd.github.drax-preview+json',
                'User-Agent': 'NodeJS'
            },
            method: 'GET'
        }, function(res) {
            let data = '';
            res.setEncoding('utf8');
            res.on('data', function (chunk) {
                data += chunk.toString();
            });

            res.on('end', function() {
                if (res.statusCode !== 200 && res.statusCode !== 302) {
                    return reject(
                        new Error(`Response status code: ${res.statusCode}. ${data}`)
                    );
                }

                return resolve(JSON.parse(data));
            });
        });

        req.once('error', reject);
        req.end();
    });
}

/**
 * @param {Object} dependencies
 * @param {Array<String>} args - optional npm install arguments
 * @param {String} cwd
 * @param {Integer} verbosity
 * @return {Promise}
 */
Template._npmInstall = function npmInstall(dependencies, args, cwd, verbosity) {
    args = _.clone(args) || [];

    return new Promise(function(resolve, reject) {
        args = args.concat(
            _.reduce(dependencies, function(out, val, key) {
                out.push(val !== '*' ? `${key}@${val}`: key);
                return out;
            }, [])
        );

        args.unshift('install');

        let npmCmd = process.platform === 'win32' ? 'npm.cmd' : 'npm';

        if (verbosity) {
            console.info(chalk.cyan('[INFO]') + ' Installing npm dependencies...');
            console.info(chalk.cyan(`[${npmCmd} ${args.join(' ')}]`));
        }

        let proc = childProcess.spawn(npmCmd, args, {cwd: cwd});

        let stderr = '';
        proc.stdout.on('data', function(data) {
            if (verbosity >= 2) {
                console.info(data.toString());
            }
        });
        proc.stderr.on('data', function(data) {
            stderr += data.toString();
            if (verbosity >= 2) {
                console.info(data.toString());
            }
        });
        proc.on('close', function(code) {
            if (code !== 0) {
                return reject(new Error(stderr));
            }

            return resolve();
        });
    });
}
